python/src/iceberg/io/s3.py

```diff
@@ -0,0 +1,185 @@
+        self._s3_client = s3_client
+        return self
+
+    def __enter__(self):
+        """Enter context for S3InputFile
+
+        This sets a botocore.response.StreamingBody instance for the
+        location to self.input_stream. If an `s3_client` is not provided through
+        a call during context open, a boto3 session using the environment AWS
+        credentials will be created.
+
+        Example:
+            >>> input_file = S3InputFile(location="s3://foo/bar.json")
+                with input_file as f:
+                    data = input_file.input_stream.read()
+
+            >>> import boto3
+                AWS_ACCESS_KEY_ID = "..."
+                AWS_SECRET_ACCESS_KEY = "..."
+                s3_client = boto3.Session(
+                    aws_access_key_id=AWS_ACCESS_KEY_ID,
+                    aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
+                ).client("s3")
+                input_file = S3InputFile(location="s3://baz/qux.json")
+                with input_file(s3_client=s3_client) as f:
+                    data = input_file.input_stream.read()
+
+
+        """
+        super().__enter__()
+        s3 = self._create_s3_session() if not self._s3_client else self._s3_client
+        s3_url = urlparse(self.location)
+        resource = s3.get_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
+        return resource["Body"]
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """A NOOP since boto3 automatically handles closing the session"""
+        super().__exit__(exc_type, exc_value, exc_traceback)
+        del self._s3_client
+        return
+
+
+class S3OutputFile(OutputFile):
+    def __init__(self, location: str):
+        super().__init__(location=location)
+        self._file_obj = None
+        self._s3_client = None
+        self._overwrite = False
+
+    def __call__(self, file_obj, overwrite: bool = False, **kwargs):
+        super().__call__(file_obj=file_obj, overwrite=overwrite)
+        self._s3_client = kwargs.get("s3_client")
+        return self
+
+    def __len__(self):
+        s3 = self._create_s3_session() if not self._s3_client else self._s3_client
+        s3_url = urlparse(self.location)
+        resource = s3.head_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
+        return resource["ContentLength"]
+
+    def _create_s3_session(self):
+        return boto3.Session().client("s3")
+
+    def exists(self, s3_client=None):
+        s3 = self._create_s3_session() if not s3_client else s3_client
+        s3_url = urlparse(self.location)
+        try:
+            s3.head_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
+        except ClientError as e:
+            if "Not Found" in str(e):
+                return False
+            else:
+                raise
+
+        return True
+
+    def to_input_file(self):
+        return S3InputFile(location=self._location)
```

Comment:
If the length was cached, you may want to pass it to the `InputFile` here.

Reply:
Great idea, updated!